packages/nemo-evaluator-launcher/src/nemo_evaluator_launcher/common/container_metadata/registries.py

```diff
@@ -0,0 +1,809 @@
+            if not www_authenticate:
+                logger.error("No WWW-Authenticate header in 401 response")
+                return False
+
+            auth_params = self._parse_www_authenticate(www_authenticate)
+            if not auth_params:
+                return False
+
+            self.bearer_token = self._request_token(auth_params, repository)
+            if not self.bearer_token:
+                return False
+
+            self.session.headers.update(
+                {
+                    "Authorization": f"Bearer {self.bearer_token}",
+                    "Accept": DOCKER_MANIFEST_MEDIA_TYPE,
+                }
+            )
+
+            return True
+
+        except Exception as e:
+            logger.error("Authentication error", error=str(e))
+            return False
+
+    # get_manifest_and_digest and get_blob are inherited from base class
+
+
+def _resolve_gitlab_credentials(
+    registry_url: str,
+) -> tuple[Optional[str], Optional[str]]:
+    """Resolve GitLab credentials from environment variables and Docker config.
+
+    Args:
+        registry_url: Registry URL
+
+    Returns:
+        Tuple of (username, password)
+    """
+    username = os.getenv("DOCKER_USERNAME")
+    password = os.getenv("GITLAB_TOKEN")
+
+    # If password from env but no username, try Docker config for username
+    if password and not username:
+        docker_creds = _read_docker_credentials(registry_url)
+        if docker_creds:
+            username, _ = docker_creds
+        else:
+            username = "gitlab-ci-token"
+
+    # If no password from env, try Docker config
+    if not password:
+        docker_creds = _read_docker_credentials(registry_url)
+        if docker_creds:
+            username, password = docker_creds
+
+    return username, password
+
+
+def _resolve_nvcr_credentials(registry_url: str) -> tuple[Optional[str], Optional[str]]:
+    """Resolve NVCR credentials from environment variables and Docker config.
+
+    Args:
+        registry_url: Registry URL
+
+    Returns:
+        Tuple of (username, password)
+    """
+    username = os.getenv("NVCR_USERNAME") or os.getenv("DOCKER_USERNAME", "$oauthtoken")
+    password = os.getenv("NVCR_PASSWORD") or os.getenv("NVCR_API_KEY")
+
+    # If no password from env, try Docker config
+    if not password:
+        docker_creds = _read_docker_credentials(registry_url)
+        if docker_creds:
+            username, password = docker_creds
+
+    # Allow None credentials for anonymous access
+    if not password:
+        username = None
+        password = None
+
+    return username, password
+
+
+def create_authenticator(
+    registry_type: str, registry_url: str, repository: Optional[str] = None
+) -> RegistryAuthenticator:
+    """Create the appropriate authenticator based on registry type.
+
+    Unified authenticator creation that supports:
+    1. Public containers (anonymous token access)
+    2. Environment variable credentials
```

Comment:
I guess this answers my question about docker-less users (they can always set env vars to unblock themselves)